packages/nimble-components/src/tree-view/index.ts

```diff
@@ -29,6 +33,16 @@ export class TreeView extends FoundationTreeView {
     @attr({ attribute: 'selection-mode' })
     public selectionMode: TreeViewSelectionMode = TreeViewSelectionMode.all;
 
+    private readonly groupSelectedItems: {
+        item: HTMLElement,
+        count: number
+    }[] = [];
```

Comment:
Maybe we need to discuss it but I'm concerned about any caches of state outside the DOM tree that hold onto node you don't own. It's not clear to me that disconnected callbacks will clear this state, etc.

The safe bet is to rely on the DOM state during these updates. If these kind of caches are helpful then maybe we can use find an alternative to use like WeakMaps

Reply:
Another way to think about it is the DOM is already a refcounted way to maintain the state. Can we just ask it qSA('[role="tree-item"][selected]) and qSA('[role="tree-item"][groupSelectedAttribute])

Reply:
I had tried that, but I found that when `onSelectedChange` is called, only the tree item's `selected` property is up-to-date. The attribute is still in the previous state. Thus `qSA` doesn't get me a current snapshot. I tried to work around this, using the knowledge that the toggled item would have the old attribute value, but that also doesn't work, because when you change the selection you get `onSelectedChange` for both the newly selected and newly deselcted items before the attributes update. So there can be a second out-of-date item, and we can't account for that.

Another thought was that I could put a selected descendant count property on the TreeItem. Then I could use the ref count logic, but not keep a cache in the TreeView. Thoughts?

Reply:
Yea fast syncs attributes asynchronously, what about qsa role="tree-item" and then just iterating across them and checking the selected property?

> Another thought was that I could put a selected descendant count property on the TreeItem. Then I could use the ref count logic, but not keep a cache in the TreeView. Thoughts?

I really don't want the ref count logic, the state of the dom should represent the visuals.

Reply:
Okay, converged on an approach that covers adding/removing elements and is less imperative. I forked the tree item template so that I could add the binding `?group-selected="${x => x.isGroupSelected()}"`. That function relies on a new property I added to both `TreeItem` and `AnchorTreeItem`: `subtreeHasSelection`. For `TreeItem`, I update that whenever: 
1. `selected` changes,
2. `childItems` changes, or
3. any child items' `subtreeHasSelection` changes

Because `AnchorTreeItem` can't have child items, I only have to keep its `subtreeHasSelection` value in sync with `selected`.